Replacing AES impls with interop

nmahendru opened the pull request:
#### 8cab381a36560866a220e47f934fa4b1bb717e0c
<pre>
Replacing AES impls with interop
<a href="https://bugs.webkit.org/show_bug.cgi?id=271680">https://bugs.webkit.org/show_bug.cgi?id=271680</a>
radar://125367961

Reviewed by Geoffrey Garen.

All the Unsafe Bits are now in UnsafeOverlays.swift.
Also all the Cpp wrappers to work around interop shortcomings are now
inside a namespace to keep them away from global scope.

Any wrappers deemed Unsafe are now inside namespace Cpp.Unsafe.

WTF::Vector Copy operation is hard fail from Swift. This is to guard
against swift&apos;s strategy of implicity copies of C++ types for exclusivity and
memory safety.

Also the quirk added in 276651@main for destructor linkage
 is moved out of the execution path.

* Source/WTF/wtf/Vector.h:
(WTF::Malloc&gt;::Vector):
* Source/WebCore/Configurations/WebCore.xcconfig:
* Source/WebCore/PAL/Configurations/PAL.xcconfig:
* Source/WebCore/PAL/pal/PALSwift.h:
(Cpp::toOptionalVectorUInt8):
(Cpp::createVectorUInt8FromSpan):
(Cpp::appendSpanToVectorUInt8):
(Cpp::Unsafe::createConstUInt8Span):
* Source/WebCore/PAL/pal/PALSwift/CryptoKitShim.swift:
(AesGcmRV.cipherText):
(AesGcmRV.errorCode):
(AesGcm.ciphertext):
(AesKwRV.result):
(AesKw.wrap(_:keyToBeWrapped:)):
(AesKw.unwrap(_:wrappedKey:)):
(AesKwRV.outputSize): Deleted.
* Source/WebCore/PAL/pal/PALSwift/UnsafeOverlays.swift:
(HashFunction.update(_:)):
(ContiguousBytes.createConstUInt8Span(_:)):
(ContiguousBytes.createConstUInt8Span):
(VectorUInt8.appendSpan(_:)):
(VectorUInt8.appendContiguousBytes(_:)):
(Data.fromSpan(_:)):
(_CompilerBugDontUse.forceLinkageForVectorDestructor):
* Source/WebCore/crypto/cocoa/CryptoAlgorithmAESGCMMac.cpp:
(WebCore::encryptCryptoKitAESGCM):
* Source/WebCore/crypto/cocoa/CryptoAlgorithmAESKWMac.cpp:
(WebCore::wrapKeyAESKWCryptoKit):
(WebCore::unwrapKeyAESKWCryptoKit):

Canonical link: <a href="https://commits.webkit.org/277285@main">https://commits.webkit.org/277285@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0bdff99ac029c0b6d8c89ad5e99a48969420f556

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/46599 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/25756 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/49206 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49276 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/42644 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/48906 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/30116 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/23218 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38000 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47178 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/22748 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/40162 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19267 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/20133 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/41278 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/4644 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/39846 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/42912 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/41657 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/51125 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/46081 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/21606 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/18008 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/45264 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/22896 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/44215 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10412 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/23338 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/53224 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/22599 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/10916 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

emw-apple commented on Source/WebCore/Configurations/WebCore.xcconfig:
```diff
@@ -35,7 +35,8 @@ WK_APPLEJPEGXL_PREPROCESSOR_DEFINITIONS_YES = USE_JPEGXL=1;
 WK_HAVE_SWIFT_CPP_INTEROP = HAVE_SWIFT_CPP_INTEROP=0;
-WK_HAVE_SWIFT_CPP_INTEROP[sdk=macos14*] = ;
+WK_HAVE_SWIFT_CPP_INTEROP[sdk=macosx14*] = ;
+WK_HAVE_SWIFT_CPP_INTEROP[sdk=macoscatalyst14*] = ;
```
`macoscatalyst` is not a thing, Catalyst uses the same `macosx` SDK.

nmahendru replied:
aah.. fixed..

sysrqb commented on Source/WebCore/PAL/pal/PALSwift.h:
```diff
@@ -30,6 +30,14 @@
+using OptionalVectorUInt8 = std::optional<WTF::Vector<uint8_t>>;
+
+// rdar://118026392
+// FIXME: remove when we stop supporting 2023 releases.
```
We don't usually reference releases by year. Maybe use something like:
```
// FIXME: Remove this when Swift support is available rdar://118026392
```

sysrqb commented on Source/WebCore/PAL/pal/PALSwift/CryptoKitShim.swift:
```diff
@@ -44,153 +45,90 @@ public enum ErrorCodes: Int {
+            sealedBox.tag[..<desiredTagLengthInBytes].forEach { val in
+                ciphertext[index] = val
+                index += 1
+            }
```
Copying these byte-by-byte into `ciphertext` is sad, especially across two loops. I'll just mention here that it'd be better if CryptoKit's SealBox could just serialize this for us.

geoffreygaren replied:
This is probably another case where we want to put an API shim inside UnsafeWrappers.swift.

sysrqb commented on Source/WebCore/PAL/pal/PALSwift/CryptoKitShim.swift:
```diff
@@ -44,153 +45,90 @@ public enum ErrorCodes: Int {
         }
-        return .success
+        return rv
+
```
extra new line

sysrqb commented on Source/WebCore/PAL/pal/PALSwift/UnsafeOverlays.swift:
```diff
@@ -37,5 +37,67 @@ extension HashFunction {
+            }
+            return rv
+        }
+
```
nit: extra new line

sysrqb commented on Source/WebCore/crypto/cocoa/CryptoAlgorithmAESGCMMac.cpp:
```diff
@@ -54,14 +54,11 @@ ALLOW_DEPRECATED_DECLARATIONS_END
 {
-    // Protect against underflow in the subtraction below.
-    if (desiredTagLengthInBytes > CryptoAlgorithmAESGCM::DefaultTagLength / 8)
+    auto rv = PAL::AesGcm::encrypt(key.span(), iv.span(), additionalData.span(), plainText.span(), desiredTagLengthInBytes);
```
Is the type of `rv` on the heap, or is that on the stack? Do we need to care about ref counting?

nmahendru replied:
swift moves back the value and then it will be cleaned up by C++.

sysrqb commented on Source/WebCore/crypto/cocoa/CryptoAlgorithmAESGCMMac.cpp:
```diff
@@ -54,14 +54,11 @@ ALLOW_DEPRECATED_DECLARATIONS_END
-    cipherText.resize(cipherText.size() - (CryptoAlgorithmAESGCM::DefaultTagLength - desiredTagLengthInBytes));
-    return WTFMove(cipherText);
+    auto ciphertext = rv.getCipherText();
+    return WTFMove(*ciphertext);
```
Can we move this directly without the intermediate `ciphertext` variable?

geoffreygaren commented on Source/WebCore/PAL/pal/PALSwift/UnsafeOverlays.swift:
```diff
@@ -37,5 +37,67 @@ extension HashFunction {
+            index += 1
+        }
+        return rv
+    }
```
The thesis of UnsafeOverlays is:
* The implementation is allowed to perform unsafe operations
* The output should be safe objects that enforce safety invariants

Consistent with that thesis, we can do something more efficient here than byte-by-byte copy:

```
return self.withUnsafeBytes { unsafeBytes in
    return WTFMove(Vector(span(unsafeBytes.baseAddress!, unsafeBytes.count)
}
```

geoffreygaren commented on Source/WebCore/PAL/pal/PALSwift/UnsafeOverlays.swift:
```diff
@@ -37,5 +37,67 @@ extension HashFunction {
+            bytesNoCopy: UnsafeMutablePointer(mutating: spanNoCopy.__dataUnsafe()),
+            count: spanNoCopy.size(), deallocator: .none)
+    }
+}
```
This abstraction is inconsistent with the thesis of UnsafeOverlays. We're returning a reference type (Data) that has a raw pointer in it, with no way to ensure the lifetime of the pointed-to object.

If there are APIs that require Data, and we have a span, probably we need to write overlays for those APIs, where the overlays accept span<T> or similar and then invoke the API directly. The overlay then does the job of encapsulating the unsafe Data object.

geoffreygaren commented on Source/WebCore/PAL/pal/PALSwift/UnsafeOverlays.swift:
```diff
@@ -37,5 +37,67 @@ extension HashFunction {
+        }
+
+    }
+}
```
Same comment here about toVectorUInt8. 

Also, if this is going to happen more than once, you can probably make a generic function that accepts any ContiguousBytes thingy, and copies to vector.

Side point: We prefer to use "to" to indicate conversion, and "copyTo" to indicate a copy. Let's call this "copyTo".

geoffreygaren commented on Source/WebCore/PAL/pal/PALSwift/UnsafeOverlays.swift:
```diff
@@ -37,5 +37,67 @@ extension HashFunction {
+        }
+        return rv
+    }
+}
```
Ditto.

geoffreygaren commented on Source/WebCore/PAL/pal/PALSwift.h:
```diff
@@ -28,8 +28,39 @@
+
+
+// FIXME: remove when swift support is available rdar://118026392
+inline OptionalVectorUInt8 toOptionalVectorUInt8(VectorUInt8 val)
```
WebKit naming convention here would be "makeOptional". We leave the type name out because it's in the argument signature. We use "make" to indicate constructing something, e.g. makeString().

nmahendru replied:
renamed.

geoffreygaren commented on Source/WebCore/PAL/pal/PALSwift.h:
```diff
@@ -28,8 +28,39 @@
+    return val;
+}
+
+inline VectorUInt8 createVectorUInt8FromSpan(SpanConstUInt8 span)
```
makeVectorUInt8.

Does the Vector(std::span<const T>) constructor not work in Swift? If so, that probably deserves a Radar and a comment linking to the Radar.

geoffreygaren commented on Source/WebCore/PAL/pal/PALSwift.h:
```diff
@@ -28,8 +28,39 @@
 #include <cstdint>
 #include <wtf/Vector.h>
 
+namespace Cpp {
```
Are we trying to resolve a namespace conflict, or something else?

nmahendru replied:
Trying to do two things here:
1. Keep all the wrapper code inside a namespace and not pollute the Global namespace.
2. Use a nested namespace `Unsafe` to keep the bits we would like to get rid of when possible.

geoffreygaren commented on Source/WebCore/PAL/pal/PALSwift.h:
```diff
@@ -28,8 +28,39 @@
+    return VectorUInt8(span);
+}
+
+inline void appendSpanToVectorUInt8(const SpanConstUInt8& span, VectorUInt8& vec)
```
I would call this append() and put the Vector argument first. It's just a style preference, but that's how most WebKit code works -- object first, argument second. 

Does Vector::append(std::span<U, Extent>) not work in Swift? If so, probably needs a Radar and comment. The purpose of the comment is to help explain the programming model. If we're doing something to work around a bug, saying so explicitly can help the reader.

nmahendru replied:
> Does Vector::append(std::span<U, Extent>) not work in Swift


Nope, cause it's templated.

geoffreygaren commented on Source/WebCore/PAL/pal/PALSwift.h:
```diff
@@ -28,8 +28,39 @@
+inline SpanConstUInt8 createConstUInt8Span(const uint8_t * start, size_t length)
+{
+    return { start, length };
+}
```
This abstraction doesn't meet our unsafe design goal because it returns the unsafe pointer in a span rather than returning a safe abstraction.

nmahendru replied:
I have left it inside Unsafe intentionally for that reason. the other way would be to never create a span from withUnsafeBytes and just always create a Vector by Copying elements in a loop.
Span constructors are not available in swift as they are templated.

geoffreygaren commented on Source/WebCore/PAL/pal/PALSwift/CryptoKitShim.swift:
```diff
@@ -44,153 +48,76 @@ public enum ErrorCodes: Int {
-                    using: symKey,
-                    nonce: nonce
-                )
+            let sealedBox: AES.GCM.SealedBox = try AES.GCM.seal(message, key: key, iv: iv, ad: ad)
```
Do you need to explicitly declare the type here or can Swift infer it?

nmahendru replied:
nope. That is probably a remnant of refactoring. Removing.

geoffreygaren commented on Source/WebCore/PAL/pal/PALSwift/CryptoKitShim.swift:
```diff
@@ -44,153 +48,76 @@ public enum ErrorCodes: Int {
+            let sealedBox: AES.GCM.SealedBox = try AES.GCM.seal(message, key: key, iv: iv, ad: ad)
+            var ciphertext: VectorUInt8 = Cpp.createVectorUInt8FromSpan(
+                sealedBox.ciphertext.createConstUInt8Span())
+            if desiredTagLengthInBytes > sealedBox.tag.count {
```
I would move this check up before the cipertext declaration. It's surprising to allocate a vector before knowing you're going to use it.

nmahendru replied:
for sure.. sorry for that!

geoffreygaren commented on Source/WebCore/PAL/pal/PALSwift/CryptoKitShim.swift:
```diff
@@ -44,153 +48,76 @@ public enum ErrorCodes: Int {
-                    nonce: nonce
-                )
+            let sealedBox: AES.GCM.SealedBox = try AES.GCM.seal(message, key: key, iv: iv, ad: ad)
+            var ciphertext: VectorUInt8 = Cpp.createVectorUInt8FromSpan(
```
I think the way to do this is to have an unsafe method on ContiguousBytes that invokes the Vector constructor. I believe that sealedBox.ciphertext conforms to ContiguousBytes. Then you can do sealedBox.ciphertext.copyToVector() or copyToVector(sealedBox.ciphertext) or similar.

geoffreygaren commented on Source/WebCore/PAL/pal/PALSwift/CryptoKitShim.swift:
```diff
@@ -44,153 +48,76 @@ public enum ErrorCodes: Int {
-            sealedBox.tag.copyBytes(
-                to: cipherText + Int(plainTextSize), count: sealedBox.tag.count)
+            ciphertext.appendSpan(
+                span: try sealedBox.tag.createConstUInt8Span(size: desiredTagLengthInBytes))
```
Given that sealedBox.tag and ciphertext are both Data, which is a safe type, there should be a way to append safely without the C++ conversion.

nmahendru replied:
I guess that's better. Doing that.

geoffreygaren commented on Source/WebCore/PAL/pal/PALSwift/UnsafeOverlays.swift:
```diff
@@ -30,12 +30,92 @@ import CryptoKit
+    }
+}
+
+// Ugly but needed.
```
Comments should try to express facts, so we can leave out this judgement.

geoffreygaren commented on Source/WebCore/PAL/pal/PALSwift/UnsafeOverlays.swift:
```diff
@@ -30,12 +30,92 @@ import CryptoKit
+// all __dataUnsafe() calls here return nil if input WTF::Vector is zero length.
+// CommonCrypto APIs handled nil pointers internally. Swift APIs crash.
+// To maintain backward compatibility, a zero array is needed to go past Swift Validation
+// and let CyptoKit replicate behavior.
```
I would say, "CryptoKit does not support a null pointer with zero length. We instead need to pass an empty Data. This class provides that."

geoffreygaren commented on Source/WebCore/PAL/pal/PALSwift/UnsafeOverlays.swift:
```diff
@@ -30,12 +30,92 @@ import CryptoKit
+// and let CyptoKit replicate behavior.
+private struct Zero {
+    public static var zeroData = Data([UInt8]())
+}
```
I would call this "Empty". That's our usual term of art.

I would make "empty()" an extension on Data. This is a way to construct an empty Data, so it belongs on the Data class.

geoffreygaren commented on Source/WebCore/PAL/pal/PALSwift/UnsafeOverlays.swift:
```diff
@@ -30,12 +30,92 @@ import CryptoKit
+}
+
+extension ContiguousBytes {
+    public func createVectorUInt8() -> VectorUInt8 {
```
I would call this copyToVector. That's just the term of art we use elsewhere.

geoffreygaren commented on Source/WebCore/PAL/pal/PALSwift/UnsafeOverlays.swift:
```diff
@@ -30,12 +30,92 @@ import CryptoKit
+extension ContiguousBytes {
+    public func createVectorUInt8() -> VectorUInt8 {
+        return self.withUnsafeBytes { buf in
+            var result = VectorUInt8(buf.count)
```
This function achieves our goal. But it is forced to name the number of bytes four times, and technically copyBytes skips the bounds check. It would be slightly nicer if you could make an explicit "makeSpan" or "makeSpanUInt8" function in this file, and then we'd have just one line of code:

`var result = VectorUInt8(makeSpan(buf.baseAddress, buf.count))`

geoffreygaren commented on Source/WebCore/PAL/pal/PALSwift/UnsafeOverlays.swift:
```diff
@@ -30,12 +30,92 @@ import CryptoKit
+}
+
+extension Data {
+    public static func fromSpan(spanNoCopy: SpanConstUInt8) -> Data {
```
We should make this fileprivate because it's not a safe concept to export. Maybe all extensions should be so?

Maybe also name this unsafeTemporaryDataFromSpan, to remind us not to return it. (We decided not to do `Data.temporaryDataFromSpan { data in ... }` because you would end up nested a few levels deep, which is not super readable.)

geoffreygaren commented on Source/WebCore/PAL/pal/PALSwift/UnsafeOverlays.swift:
```diff
@@ -30,12 +30,92 @@ import CryptoKit
     }
 }
 
+extension SpanConstUInt8: ContiguousBytes {
```
fileprivate

geoffreygaren commented on Source/WebCore/PAL/pal/PALSwift/UnsafeOverlays.swift:
```diff
@@ -30,12 +30,92 @@ import CryptoKit
+    }
+}
+
+private class _CompilerBugDontUse {
```
I would rename this to _WorkAroundRdar116406681.

geoffreygaren reviewed (approved):
r=me if you make the fileprivate changes, and the other comments are nice to have if they work out, but don't block landing.

emw-apple commented on Source/WTF/wtf/Vector.h:
```diff
@@ -1026,6 +1026,9 @@ template<typename T, size_t inlineCapacity, typename OverflowHandler, size_t min
     : Base(other.size(), other.size())
 {
+#ifdef __swift__
+    RELEASE_ASSERT_NOT_REACHED_WITH_MESSAGE("C++ copy constructors are not memory safe in Swift. Try to find a way to borrow the object instead.");
```
I'm a little impressed that this works at all—that Swift will import an inline definition and generate code for it—but I think the C++ model relies on the possibility of multiple compilation units generating the same function, and the linker being allowed to pick one arbitrarily.

i.e. if Swift instantiates a `Vector<P>` but some C++ code also uses a `Vector<P>`, both compilation units will generate the same constructor, but only the former will have the release assertion. And the linker makes no guarantee about which one it'll use, but there will only be _one_ for the whole binary.

emw-apple replied:
You might be able to enforce this at compile time with `std::enable_if` or a `static_assert`.

emw-apple replied:
To give you a more rigorous answer, yes, this appears to violate the [One Definition Rule](https://en.cppreference.com/w/cpp/language/definition):

> There can be more than one definition in a program of each of the following: class type, enumeration type, inline function, inline variable(since C++17), [templated entity](https://en.cppreference.com/w/cpp/language/templates#Templated_entity) (template or member of template, but not full [template specialization](https://en.cppreference.com/w/cpp/language/template_specialization)), as long as all of the following is true:
> - each definition appears in a different translation unit
> …
> - **each definition consists of the same sequence of tokens** (typically, appears in the same header file)
> …
> 
> If all these requirements are satisfied, the program behaves as if there is only one definition in the entire program. Otherwise, the program is ill-formed, no diagnostic required.

nmahendru committed:
Replacing AES impls with interop
https://bugs.webkit.org/show_bug.cgi?id=271680
radar://125367961

Reviewed by Geoffrey Garen.

All the Unsafe Bits are now in UnsafeOverlays.swift.
Also all the Cpp wrappers to work around interop shortcomings are now
inside a namespace to keep them away from global scope.

Any wrappers deemed Unsafe are now inside namespace Cpp.Unsafe.

WTF::Vector Copy operation is hard fail from Swift. This is to guard
against swift's strategy of implicity copies of C++ types for exclusivity and
memory safety.

Also the quirk added in 276651@main for destructor linkage
 is moved out of the execution path.

* Source/WTF/wtf/Vector.h:
(WTF::Malloc>::Vector):
* Source/WebCore/Configurations/WebCore.xcconfig:
* Source/WebCore/PAL/Configurations/PAL.xcconfig:
* Source/WebCore/PAL/pal/PALSwift.h:
(Cpp::toOptionalVectorUInt8):
(Cpp::createVectorUInt8FromSpan):
(Cpp::appendSpanToVectorUInt8):
(Cpp::Unsafe::createConstUInt8Span):
* Source/WebCore/PAL/pal/PALSwift/CryptoKitShim.swift:
(AesGcmRV.cipherText):
(AesGcmRV.errorCode):
(AesGcm.ciphertext):
(AesKwRV.result):
(AesKw.wrap(_:keyToBeWrapped:)):
(AesKw.unwrap(_:wrappedKey:)):
(AesKwRV.outputSize): Deleted.
* Source/WebCore/PAL/pal/PALSwift/UnsafeOverlays.swift:
(HashFunction.update(_:)):
(ContiguousBytes.createConstUInt8Span(_:)):
(ContiguousBytes.createConstUInt8Span):
(VectorUInt8.appendSpan(_:)):
(VectorUInt8.appendContiguousBytes(_:)):
(Data.fromSpan(_:)):
(_CompilerBugDontUse.forceLinkageForVectorDestructor):
* Source/WebCore/crypto/cocoa/CryptoAlgorithmAESGCMMac.cpp:
(WebCore::encryptCryptoKitAESGCM):
* Source/WebCore/crypto/cocoa/CryptoAlgorithmAESKWMac.cpp:
(WebCore::wrapKeyAESKWCryptoKit):
(WebCore::unwrapKeyAESKWCryptoKit):

Canonical link: https://commits.webkit.org/277285@main
